softlayer/resource_softlayer_file_storage.go

```diff
@@ -0,0 +1,941 @@
+	case PerformanceType:
+		return fmt.Sprintf("%.f_IOPS", iops), nil
+	}
+	return "", fmt.Errorf("Invalid storageType %s.", storageType)
+}
+
+func getPrice(productItems []datatypes.Product_Item, keyName string, categoryCode string, capacityRestrictionType string, capacityRestriction int) (datatypes.Product_Item_Price, error) {
+	for _, item := range productItems {
+		if strings.HasPrefix(*item.KeyName, keyName) {
+			for _, price := range item.Prices {
+				if *price.Categories[0].CategoryCode == categoryCode && price.LocationGroupId == nil {
+					if capacityRestrictionType == "STORAGE_SPACE" {
+						if price.CapacityRestrictionMinimum == nil ||
+							price.CapacityRestrictionMaximum == nil {
+							continue
+						}
+						capacityRestrictionMinimum, _ := strconv.Atoi(*price.CapacityRestrictionMinimum)
+						capacityRestrictionMaximum, _ := strconv.Atoi(*price.CapacityRestrictionMaximum)
+						if capacityRestrictionMinimum > 0 &&
+							capacityRestriction >= capacityRestrictionMinimum &&
+							capacityRestriction <= capacityRestrictionMaximum {
+							return price, nil
+						}
+					}
+
+					if capacityRestrictionType == "STORAGE_TIER_LEVEL" {
+						if price.CapacityRestrictionMinimum == nil ||
+							price.CapacityRestrictionMaximum == nil {
+							continue
+						}
+						capacityRestrictionMinimum, _ := strconv.Atoi(*price.CapacityRestrictionMinimum)
+						capacityRestrictionMaximum, _ := strconv.Atoi(*price.CapacityRestrictionMaximum)
+						if capacityRestrictionMinimum > 0 &&
+							capacityRestriction == capacityRestrictionMinimum &&
+							capacityRestriction == capacityRestrictionMaximum {
+							return price, nil
+						}
+					}
+
+					if capacityRestrictionType == "" && capacityRestriction == 0 {
+						return price, nil
+					}
+				}
+			}
+		}
+	}
+	return datatypes.Product_Item_Price{},
+		fmt.Errorf("No product items matching with keyName %s and categoryCode %s could be found", keyName, categoryCode)
+}
+
+func getIops(storage datatypes.Network_Storage, storageType string) (float64, error) {
+	switch storageType {
+	case EnduranceType:
+		for _, property := range storage.Properties {
+			if *property.Type.Keyname == "PROVISIONED_IOPS" {
+				provisionedIops, err := strconv.Atoi(*property.Value)
+				if err != nil {
+					return 0, err
+				}
+				return float64(provisionedIops) / float64(*storage.CapacityGb), nil
+			}
+		}
+	case PerformanceType:
+		if storage.Iops == nil {
+			return 0, fmt.Errorf("Failed to retrive iops information.")
+		}
+		iops, err := strconv.Atoi(*storage.Iops)
+		if err != nil {
+			return 0, err
+		}
+		return float64(iops), nil
+	}
+	return 0, fmt.Errorf("Invalied storage type %s", storageType)
+}
+
+func updateAllowedIpAddresses(d *schema.ResourceData, sess *session.Session, storage datatypes.Network_Storage) error {
+	id := *storage.Id
+	newIps := d.Get("allowed_ip_addresses").(*schema.Set).List()
+
+	// Add new allowed_ip_addresses
+	for _, newIp := range newIps {
```

Comment:
This logic should be extracted to another function and unit tested. The rest of this function is calls out to a server and is better covered by the acceptance tests.